src/infer.c

```diff
@@ -0,0 +1,333 @@
+* limitations under the License.
+*/
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "infer.h"
+#include "color.h"
+#include "maps.h"
+#include "util.h"
+
+/* Uses directly in the testsuite */
+enum selint_error infer_interfaces_shallow(const struct policy_node *node);
+enum selint_error infer_interfaces_deep(const struct policy_node *node);
+
+static enum param_flavor name_to_param_flavor(enum name_flavor flavor)
+{
+	switch (flavor) {
+	case NAME_TYPE:
+		return PARAM_TYPE;
+	case NAME_TYPEATTRIBUTE:
+		return PARAM_TYPEATTRIBUTE;
+	case NAME_TYPE_OR_ATTRIBUTE:
+		return PARAM_TYPE_OR_ATTRIBUTE;
+	case NAME_ROLE:
+		return PARAM_ROLE;
+	case NAME_ROLEATTRIBUTE:
+		return PARAM_ROLEATTRIBUTE;
+	case NAME_ROLE_OR_ATTRIBUTE:
+		return PARAM_ROLE_OR_ATTRIBUTE;
+	case NAME_CLASS:
+		return PARAM_CLASS;
+	case NAME_OBJECT_NAME:
+		return PARAM_OBJECT_NAME;
+	default:
+		// should never happen
+		return PARAM_UNKNOWN;
+	}
+}
+
+enum infer_type { IN_SHALLOW, IN_DEEP };
+
+struct infer_data {
+	struct interface_trait *if_data;
+	enum infer_type mode;
+	const struct policy_node *node;
+};
+
+static const char *trait_type_to_str(enum trait_type t)
+{
+	switch (t) {
+	case INTERFACE_TRAIT:
+		return "interface";
+	case TEMPLATE_TRAIT:
+		return "template";
+	case MACRO_TRAIT:
+		return "macro";
+	default:
+		// should never happen
+		return "unknown-trait-type";
+	}
+}
+
+static void infer_func(const char *name, enum name_flavor flavor, unsigned short id, void *visitor_data)
+{
+	if (!name) {
+		return;
+	}
+
+	struct infer_data *data = visitor_data;
+
+	const char *dollar = strchr(name, '$');
+	if (!dollar) {
+		return;
+	}
+
+	if (0 == strcmp(name, "$*") && flavor == NAME_IF_PARAM && id == 1) {
+		const struct interface_trait *call_trait = look_up_in_if_traits_map(data->node->data.ic_data->name);
+		if (!call_trait) {
+			print_if_verbose("No call trait for %s\n", data->node->data.ic_data->name);
+		} else {
+			for (int i = 0; i < TRAIT_MAX_PARAMETERS; i++) {
+				data->if_data->parameters[i] = call_trait->parameters[i];
+			}
+		}
+		return;
+	}
+
+	char *param_end;
+	errno = 0;
+	unsigned long param_no = strtoul(dollar + 1, &param_end, 10);
+	if (param_no == 0 || errno != 0) {
+		fprintf(stderr, "%sError%s: Failed to parse parameter number from name '%s' in %s %s!\n",
+				color_error(), color_reset(),
+				name,
+				trait_type_to_str(data->if_data->type),
+				data->if_data->name);
+		return;
+	}
+	param_no--; // start counting at 0 ($0 is invalid)
+	if (param_no > TRAIT_MAX_PARAMETERS) {
+		fprintf(stderr, "%sWarning%s: Only up to %u parameters supported, parsed %lu from name '%s' in %s %s!\n",
+				color_warning(), color_reset(),
+				TRAIT_MAX_PARAMETERS,
+				param_no,
+				name,
+				trait_type_to_str(data->if_data->type),
+				data->if_data->name);
+		return;
+	}
+
+	// skip dash of exclusions
+	if (name[0] == '-') {
+		name = name + 1;
+	}
+
+	if (dollar == name && *param_end == '\0') {
```

Comment:
I may very well be missing something here, but I'm concerned about the situation where a parameter is incorrectly used in a way that would infer as two different parameters.  Something like:

allow $1 foo_t:file read;
role $1 types foo_t;

This would be a compile error, but I *think* the logic here would end up setting the flavor of $1 to PARAM_ROLE, since the second line just overwrites the first?  It seems like it would be valuable info to have to know if a parameter was being used inconsistently.

Then again, I could definitely be missing something, so feel free to let me know if this scenario is already handled.

Reply:
True, a subsequent usage of the same no yet final inferred parameter will completely override the current type.
Probably the inferred state shouldn't be a type of an enum but a bit-mask where its possible to store the information a parameter is used as a type-or-attribute, a role and a template string-part. So analysis checks can then create reports like "Parameter X has conflicting usages", "Interface foo uses templated parameter X (declare as template)".